### ISSUE TITLE
MX3 vertical scrolling becomes slow, probably after suspenc

### ISSUE THREAD

**Issue by janvlug** [pwr-Solaar/Solaar#951]
**Information**

- Solaar version: 1.0.3
- Distribution: Fedora 32
- Kernel version: `Linux 5.8.8-200.fc32.x86_64 x86_64 GNU/Linux`
- Output of `solaar show`:
```
Unifying Receiver
  Device path  : /dev/hidraw0
  USB id       : 046d:c52b
  Serial       : B0E8918D
    Firmware   : 12.11.B0032
    Bootloader : 04.16
    Other      : AA.AA
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless (0x000100)
  Device activity counters: 1=37

  1: Wireless Mouse MX Master 3
     Codename     : MX Master 3
     Kind         : mouse
     Wireless PID : 4082
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: 50A2AF6F
        Bootloader: BOT 95.01.B0015
          Firmware: MPM 19.01.B0015
             Other: 
     The power switch is located on the base.
     Supports 35 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: CRYPTO ID              {0021}   
         7: DEVICE FRIENDLY NAME   {0007}   
         8: BATTERY STATUS         {1000}   
         9: REPROG CONTROLS V4     {1B04}   
            Actions: {'80': 80, '81': 81, '82': 82, '83': 83, '86': 86, '195': 195, '196': 196}
        10: CHANGE HOST            {1814}   
            Change Host: 2:nyx
        11: unknown:2250           {2250}   
        12: ADJUSTABLE DPI         {2201}   
            Sensitivity (DPI): 1000
        13: SMART SHIFT            {2110}   
            Smart Shift: 0
        14: HIRES WHEEL            {2121}   
            Multiplier: 15
            Has invert
              Normal wheel motion
            Has ratchet switch
              Free wheel mode
            Low resolution mode
            HID notification
            High Resolution Wheel Invert: False
            Wheel Resolution: False
        15: THUMB WHEEL            {2150}   
            HID++ Thumb Scrolling: False
            Thumb Scroll Invert: False
        16: unknown:2251           {2251}   
        17: DFUCONTROL SIGNED      {00C2}   
        18: DEVICE RESET           {1802}   internal, hidden
        19: unknown:1803           {1803}   internal, hidden
        20: CONFIG DEVICE PROPS    {1806}   internal, hidden
        21: unknown:1813           {1813}   internal, hidden
        22: OOBSTATE               {1805}   internal, hidden
        23: unknown:1830           {1830}   internal, hidden
        24: unknown:1890           {1890}   internal, hidden
        25: unknown:1891           {1891}   internal, hidden
        26: unknown:18A1           {18A1}   internal, hidden
        27: unknown:1DF3           {1DF3}   internal, hidden
        28: unknown:1E00           {1E00}   hidden
        29: unknown:1EB0           {1EB0}   internal, hidden
        30: unknown:1861           {1861}   internal, hidden
        31: unknown:9001           {9001}   internal, hidden
        32: unknown:9203           {9203}   internal, hidden
        33: unknown:9205           {9205}   internal, hidden
        34: unknown:9300           {9300}   internal, hidden
     Has 8 reprogrammable keys:
         0: Left Button               , default: Left Click                  => Left Button               
             mse, pos:0, group:1, gmask:1
         1: Right Button              , default: Right Click                 => Right Button              
             mse, pos:0, group:1, gmask:1
         2: Middle Button             , default: Mouse Middle Button         => Middle Button             
             mse, reprogrammable, divertable, pos:0, group:3, gmask:7
         3: Back Button               , default: Mouse Back Button           => Back Button               
             mse, reprogrammable, divertable, pos:0, group:2, gmask:3
         4: Forward Button            , default: Mouse Forward Button        => Forward Button            
             mse, reprogrammable, divertable, pos:0, group:2, gmask:3
         5: App Switch Gesture        , default: Gesture Button Navigation   => App Switch Gesture        
             mse, reprogrammable, divertable, pos:0, group:3, gmask:7
         6: Smart Shift               , default: Smart Shift                 => Smart Shift               
             mse, reprogrammable, divertable, pos:0, group:3, gmask:7
         7: Virtual Gesture Button    , default: Virtual Gesture Button      => Virtual Gesture Button    
             divertable, virtual, pos:0, group:4, gmask:0
     Battery: 50%, discharging, next level 20%.

EX100 Receiver 27 Mhz
  Device path  : /dev/hidraw3
  USB id       : 046d:c517
  Serial       : None
  Has 2 paired device(s) out of a maximum of 4.
  Notifications: wireless (0x000100)

  1: Wireless Mouse EX100
     Codename     : EX100m
     Kind         : mouse
     Wireless PID : 3F00
     Protocol     : HID++ 1.0
     Serial number: ?
     The power switch is located on the (unknown).
     Notifications: (none).
     Features: special buttons, scroll accel (0x420000)
     Battery: good, discharging.

  3: Wireless Keyboard EX100
     Codename     : EX100
     Kind         : keyboard
     Wireless PID : 6500
     Protocol     : HID++ 1.0
     Serial number: ?
     The power switch is located on the (unknown).
     Notifications: (none).
     Features: special buttons, scroll accel (0x420000)
     Battery: full, discharging.

```

**Describe the bug**
I have an MX Master 3. I use the scroll wheel always in "free" mode. I like to be able to scroll vertically fast.
Sometimes (I think, but am not sure, after suspend of my system) vertical scrolling becomes slow.
The Wheel Resolution setting is always on in the Solaar GUI.
To "fix" the slow vertical scrolling, I disable the Wheel Resolution setting and enable it again.

**To Reproduce**
1. Enable Wheel Resolution in Solaar settings for MX Master 3
2. Let the system and/ or mouse go to suspend mode
3. Vertical scrolling is now very slow

**Additional context**
I have the MX Master 3 connected to two Linux systems. On one using Solaar (this is the system where I'm having this issue), on the other one I use Bluetooth. I did not encounter the issue on the system where I use the mouse with Bluetooth.

**Comment by pfps** [comment]
What happens when the Wheel Resolution setting is off?  (When it is off for my MX Master 3, which appears to be exactly the same as yours, the scroll wheel appears to have no effect.)  

What might be happening is that Solaar is applying the wrong setting when the device reconnects.  This is controlled by the file ~/.config/solaar/config.json 
What you could do is remove this file and see what happens.

Solaar should be applying the settings from this file every time the mouse connects.  Either there is the wrong information in this file or Solaar is not applying the information correctly.

**Comment by thilakm-git** [comment]
Unfortunately removing that file did not solve it for me. I am experiencing the exact same with the MX Master 2S. The only workaround I have found so far is re-enabling the "wheel resolution" toggle after a reconnection.

**Comment by pfps** [comment]
Please provide output of `solaar show` when the problem occurs and then after you re-enable wheel resolution.

**Comment by janvlug** [comment]
After resume from suspend when scrolling is slow: 
[slow_scrolling.txt](https://github.com/pwr-Solaar/Solaar/files/5469024/slow_scrolling.txt)
After disabling and re-enabling the "Wheel Resolution" toggle scrolling is fast again:
[fast_scrolling.txt](https://github.com/pwr-Solaar/Solaar/files/5469025/fast_scrolling.txt)

```
$ diff slow_scrolling.txt fast_scrolling.txt 
10c10
<   Device activity counters: 1=252
---
>   Device activity counters: 1=10
48c48
<             Low resolution mode
---
>             High resolution mode
51c51
<             Wheel Resolution: False
---
>             Wheel Resolution: True
```

**Comment by pfps** [comment]
It certainly looks as if the settings are not being correctly applied after resuming.  To verify I'll need the output of `bin/solaar -dd` around the time that the problem occurs.  There should be lines containing
```
logitech_receiver.settings: hires-smooth-invert: apply 
```
and
```
logitech_receiver.settings: hires-smooth-resolution: apply
```
Please include the lines from about 20 lines before the first of these and 20 lines after the last.

**Comment by thilakm-git** [comment]
> Please provide output of `solaar show` when the problem occurs and then after you re-enable wheel resolution.

1. [normal.txt](https://github.com/pwr-Solaar/Solaar/files/5470846/normal.txt) shows the device working as intendend.
2. [slow.txt](https://github.com/pwr-Solaar/Solaar/files/5470849/slow.txt) shows the device with the slow scrolling speed, this time triggered by turning off & on the mouse and thus re-connecting to Ubuntu 20.10 on Solaar 1.0.4.

I'll provide the other requested output soon. Thanks!

**Comment by janvlug** [comment]
The following log file contains logging after that my computer resumed from suspense at 21:19.
After resuming from suspend, I used solaar to turn off and on the wheel resolution, to make vertical scrolling fast again:
[slow_after_suspend.txt](https://github.com/pwr-Solaar/Solaar/files/5483869/slow_after_suspend.txt)

**Comment by pfps** [comment]
Thanks.  Unfortunately that appears to be the output with `-d`, not `-dd`.   It is not going to be enough to completely diagnose the problem.  Please provide output with `-dd`.
 But something is definitely going wrong.   Solaar should be pushing settings to the mouse as soon as the mouse wakes up.  It may be that the problem is the other receiver connected.  Can you run solaar, suspend, and resume all without the other receiver?

**Comment by thilakm-git** [comment]
I'm not sure whether I've done it right since the output is quite little. Do let me know if I should do something differently.

While Ubuntu is running, turning off & on the mouse to trigger the slow scrolling (same as after resuming from suspense):
```
07:54:16,591     INFO [MainThread] root: language en_GB (UTF-8), translations path None
Solaar depends on a udev file that is not present
For more information see the Solaar installation directions
at https://pwr-solaar.github.io/Solaar/installation
07:54:16,689    DEBUG [MainThread] solaar.ui.tray: using AppIndicator3
07:54:16,699     INFO [MainThread] solaar.upower: connected to system dbus, watching for suspend/resume events
```
Using solaar to turn off & on scroll wheel resolution to regain intended scrolling speed.
```
07:54:36,893     INFO [MainThread] root: language en_GB (UTF-8), translations path None
Solaar depends on a udev file that is not present
For more information see the Solaar installation directions
at https://pwr-solaar.github.io/Solaar/installation
07:54:37,002    DEBUG [MainThread] solaar.ui.tray: using AppIndicator3
07:54:37,013     INFO [MainThread] solaar.upower: connected to system dbus, watching for suspend/resume events
```

**Comment by pfps** [comment]
When the mouse is turned on there should be quite a bit of output from `bin/solaar -dd`, including something like:

```
11:59:10,770    DEBUG [ReceiverListener:hidraw2] logitech_receiver.base: (19) => r[10 04 4104 925540]
11:59:10,780     INFO [ReceiverListener:hidraw2] solaar.listener: connection Notification(10,4,41,04,925540) for <Device(4,4055,Wireless Mouse M185/M235/M310,C8DC66F5)> (mouse)
11:59:10,781     INFO [ReceiverListener:hidraw2] solaar.listener: status_changed <UnifyingReceiver(/dev/hidraw2,19)>: present, 4 paired devices. (0) 
11:59:10,781    DEBUG [ReceiverListener:hidraw2] logitech_receiver.notifications: <Device(4,4055,Wireless Mouse M185/M235/M310,C8DC66F5)>: eQUAD step 4 DJ connection notification: software=True, encrypted=False, link=True, payload=True
11:59:10,781    DEBUG [ReceiverListener:hidraw2] logitech_receiver.status: <Device(4,4055,Wireless Mouse M185/M235/M310,C8DC66F5)> pushing device settings [<Setting([feature:toggle] M185/M235/M310:lowres-smooth-scroll=False)>, <Setting([feature:range] M185/M235/M310:pointer_speed=256)>, <Setting([feature:map choice] M185/M235/M310:reprogrammable-keys={'80': 80, '81': 81, '82': 82})>]
11:59:10,781    DEBUG [ReceiverListener:hidraw2] logitech_receiver.settings: lowres-smooth-scroll: apply False (<Device(4,4055,Wireless Mouse M185/M235/M310,C8DC66F5)>)
11:59:10,782    DEBUG [ReceiverListener:hidraw2] logitech_receiver.settings: lowres-smooth-scroll: settings read False from <Device(4,4055,Wireless Mouse M185/M235/M310,C8DC66F5)>
```

What is happening here is that the receiver is sending a connection notification for the mouse and Solaar is responding by checking settings for the mouse and making sure that these settings are sent to the mouse.  I'm not seeing any of this.

So please check that you are running `bin/solaar -dd` and aren't seeing anything like that.

**Comment by kokosowy** [comment]
I can confirm issue from my side. I have M720 mouse connected via unifying receiver. After waking notebook up from sleep the wheel scrolls pages just a bit like there is a problem with scroll sensitivity. After unifying receiver removal and putting it back again scroll works fine, scrolls pages more and faster. Let me know in case you need any other information/log from my side.

**Comment by pfps** [comment]
Something appears to be going wrong when your mouse wakes up.  What should happen is that a notification is being sent to Solaar and Solaar responds by changing the mouse feature settings to be the ones that Solaar remembers for the mouse.  What I need to see is the details of that interaction, which should be produced by running `solaar -dd`.
This should produce lots of output, including lines that show the I/O with the receiver.

So, run `solaar -dd`, turn your mouse off and on again, and attach the output starting from just before you turned your mouse.  You should be running version 1.0.3 or 1.0.4 of Solaar.

**Comment by janvlug** [comment]
I made a -dd log. I was 65 MB, and I had to make it smaller than 10 MB to upload here, but I still have the original if needed.
Here is the version where I removed lines to make it smaller:
[solaar-dd-reduced.txt](https://github.com/pwr-Solaar/Solaar/files/5580028/solaar-dd-reduced.txt)
The system was around 18:00 suspended (automatically after not using it for a while).
At 19:25, after resuming, I noticed that vertical scrolling became slow.
At 19:26 I disabled vertical scrolling and enabled it again, which made vertical scrolling fast again.

**Comment by pfps** [comment]
Here is what I think is going on.   Your device connects before Solaar starts up after the suspend (maybe because you are moving it).  Solaar then doesn't notice that the device might have been disconnected and doesn't push settings to the device.

This problem has been fixed in version 1.0.4 so please update to at least that version.

**Comment by janvlug** [comment]
I am using version 1.0.4 (see line 201 in the file in my previous comment), also dnf shows:
```
# dnf list installed | grep solaar
solaar.noarch                                      1.0.4-2.fc33                        @updates                                       
solaar-udev.noarch                                 1.0.4-2.fc33                        @updates
```

**Comment by pfps** [comment]
Hmm.   Then I'm again stumped.

**Comment by pfps** [comment]
Are you waking up your computer by moving the mouse?

**Comment by janvlug** [comment]
I'm not waking up the computer by moving the mouse. I will try to determine if there is a difference when I am moving the mouse when resuming, and when I resume without using the mouse.

**Comment by pfps** [comment]
The code for pushing device settings was not working correctly if the device was active on resume.   PR #1012 should fix this problem.  Please test it out.  You only need to run with `-d` as the logging has been changed to show the settings push with this flag.

Note that there have been some significant changes to Solaar so you may have to do some of the extra actions in https://pwr-solaar.github.io/Solaar/installation if there are errors when trying to run Solaar.


To download and work with Solar
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1012, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1012/head:pull_1012
git checkout pull_1012
```
To download a new version of the pull request, fetch it and then set your pull branch
to the new fetch, as in:
```
git checkout pull_1012
git fetch origin pull/1012/head
git reset --hard FETCH_HEAD
```

**Comment by BachoSeven** [comment]
A question not about the issue but about support for the MX Master 3: do all things supported for the MX master 2S work reasonably with the MX3? There's nothing about the MX3 on the partial support device capabilities page, so I was wondering if gestures and smooth/side scrolling work in anybody's experience

**Comment by pfps** [comment]
Different Logitech devices have different features, so the 3 and the 2S may behave differently.  Even what looks to be the same capability may be a different feature underneath.  Solaar can query most devices to see what features they implement so Solaar doesn't need to know about the device to be able to support its features.

I have a 3.  Solaar can adjust vertical scrolling direction and resolution, can scale movement, can change button behaviour in limited ways, and change the direction of scrolling for the thumb wheel.  A very recent extension to Solaar can adjust thumb scrolling by user-specified rules.

**Comment by joaoportela** [comment]
I'm having this same issue. 

I was using the unifying receiver without issues.
* I put the computer to sleep and left the receiver at home.
* I connected the mouse using bluetooth and the scroll was super slow. Only moved the page when I scrolled really fast.
* Enabling `hires-smooth-resolution` fixes it.
![image](https://github.com/pwr-Solaar/Solaar/assets/217176/9655b788-a43c-4488-91f8-7fc48a9fff64)
* If I disable hires again it reverts to the broken state.
* Disconnecting and connecting the mouse keeps the same behavior.

* I'm using solaar 1.1.12

Anything else that I should try?
